pkg/remotes/docker/artifacts.go

```diff
@@ -0,0 +1,266 @@
+package docker
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	remoteserrors "github.com/containerd/containerd/remotes/errors"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+func (d *dockerDiscoverer) Pusher(ctx context.Context, ref string) (remotes.Pusher, error) {
+	d.reference = ref
+	return d, nil
+}
+
+func (d *dockerDiscoverer) Push(ctx context.Context, desc ocispec.Descriptor) (content.Writer, error) {
+	switch desc.MediaType {
+	case artifactspec.MediaTypeArtifactManifest:
+		h, err := d.filterHosts(docker.HostCapabilityPush)
```

Comment:
Why do you need to filter host here?

Reply:
The host needs to be filtered here to find a host with the ability to push.